File: src/main/java/io/tripled/social/client/infrastructure/InMemoryMessages.java
package io.tripled.social.client.infrastructure;

import io.tripled.social.client.domain.Message;
import io.tripled.social.client.domain.Messages;
import io.tripled.social.client.domain.UserName;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class InMemoryMessages implements Messages {

  private List<Message> messages = new ArrayList<>();

  @Override
  public List<Message> findAll() {
    return Collections.unmodifiableList(messages);
  }

  @Override
  public void save(Message message) {
    this.messages.add(message);
  }

  @Override
  public List<Message> findAllByUserName(UserName userName) {
    return messages.stream()
        .filter(m -> m.writtenBy(userName))
        .sorted()
        .collect(Collectors.toList());
  }

  @Override
  public List<Message> findAllByUserNames(List<UserName> userNames) {
    return messages.stream()
        .filter(m -> {
          for (UserName userName : userNames) {
            if (m.writtenBy(userName)) {
              return true;
            }
          }
          return false;
        })
        .sorted()
        .collect(Collectors.toList());
  }
}

File: src/main/kotlin/io/tripled/social/client/infrastructure/InMemoryMessages.kt
package io.tripled.social.client.infrastructure

import io.tripled.social.client.domain.Message
import io.tripled.social.client.domain.Messages
import io.tripled.social.client.domain.UserName
import java.util.*
import java.util.stream.Collectors

class InMemoryMessages : Messages {
    private val messages: MutableList<Message> = ArrayList<Message>()

    override fun findAll(): List<Message> {
        return Collections.unmodifiableList<Message?>(messages)
    }

    override fun save(message: Message) {
        this.messages.add(message)
    }

    override fun findAllByUserName(userName: UserName): MutableList<Message> {
        return messages.stream()
            .filter { m: Message? -> m!!.isWrittenBy(userName) }
            .sorted()
            .collect(Collectors.toList())
    }

    override fun findAllByUserNames(userNames: List<UserName>): List<Message> {
        return messages.stream()
            .filter { m: Message? ->
                for (userName in userNames) {
                    if (m!!.isWrittenBy(userName)) {
                        return@filter true
                    }
                }
                false
            }
            .sorted()
            .collect(Collectors.toList())
    }
}
